src/matrix.c

```diff
@@ -0,0 +1,289 @@
+  { 
+    // Determine which columns in DT we're iterating through on this thread
+    int threadnum = omp_get_thread_num();
+    R_len_t startcol = threadnum * chunksize;
+    R_len_t endcol = startcol + chunksize - 1;
+    if (endcol >= dtncol) 
+      endcol = dtncol - 1;
+    
+    // Determine where to fill in the matrix vector
+    R_xlen_t vecIdx = startcol * n;
+    if (rncolnum < startcol) 
+      vecIdx -= n;
+    
+    // Iterate through columns in DT, copying the contents to the matrix column
+    for (R_len_t jj = startcol; jj <= endcol; jj++) {
+      if (jj == rncolnum)  // skip rownames.
+        continue; 
+      memcpy(pmat + vecIdx, pcol[jj], sizeof(Rcomplex)*n);
+      vecIdx += n;
+    }
+  }
+
+UNPROTECT(1);
+return mat;
+}
+
+SEXP asmatrix_character(SEXP dt, R_xlen_t matlen, R_len_t n, R_len_t rncolnum) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(STRSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Determine number of columns and rows
+  R_len_t dtncol = length(dt);
+
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_len_t jj = 0; jj < dtncol; jj++) {
+    if (jj == rncolnum) // skip rownames.
+      continue; 
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_len_t ii = 0; ii < n; ii++) {
+      SET_STRING_ELT(mat, vecIdx, STRING_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+SEXP asmatrix_list(SEXP dt, R_xlen_t matlen, R_len_t n, R_len_t rncolnum) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(VECSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Determine number of columns and rows
+  R_len_t dtncol = length(dt);
+
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_len_t jj = 0; jj < dtncol; jj++) {
+    if (jj == rncolnum) // skip rownames.
+      continue; 
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_len_t ii = 0; ii < n; ii++) {
+      SET_VECTOR_ELT(mat, vecIdx, VECTOR_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+// Dispatch function for different atomic types
+SEXP asmatrix(SEXP dt, SEXP rownames) {
+  // Determine the number of rows and columns in the data.table
+  R_xlen_t p = xlength(dt);
+  R_xlen_t n = xlength(VECTOR_ELT(dt, 0));
+  
+  /* Determine the column number in which the rownames is stored 
+   * to skip when filling in the matrix. If there are no rownames, this
+   * number will be larger than the number of columns in the data.table.
+   */
+  R_xlen_t rncolnum = asInteger(rownames) - 1; // -1 to convert from 1-based to 0-based indexing
+  if (rncolnum < p)
+    p--;
+
+  // Check neither rows or columns are greater than 2^31 - 1
+  if (((p >> 31) - 1) > 0) 
```

Comment:
I believe we should just be using `INT_MAX` macro at C level, you can see definition of `.Machine$integer.max` as `INT_MAX` here

https://github.com/wch/r-source/blob/trunk/src/main/platform.c#L149-L150

Reply:
I've updated the check and the error message.

Reply:
Rinternals.h provides macros `R_SHORT_LEN_MAX ` and `R_XLEN_T_MAX`. Should we use  `R_SHORT_LEN_MAX ` instead of `INT_MAX`?